### ISSUE TITLE
How to configure ciphers?

### ISSUE THREAD

**Issue by utelle** [utelle/SQLite3MultipleCiphers#1]
In addition to a C interface the former _wxSQLite3 encryption extension_ allowed to configure the supported ciphers via user-defined functions. For example:

```SQL
-- Get cipher used for the next key or rekey operation
SELECT wxsqlite3_config('cipher');
```

In principle, the new SQLite VFS based approach would allow to implement specialized `PRAGMA` statements. However, these `PRAGMA` statements have a serious restriction: they work only on file-based databases and not on databases held in memory, that is, a database opened with filename `:memory:`. In the latter case such `PRAGMA` statements would be silently ignored.

One could argue that for databases held in memory this doesn't impose a real problem, because they aren't encrypted anyway. However, if one attaches a file-based database to such a memory database connection, it would be necessary to use the schema prefix of the attached database, so that `PRAGMA` statements take effect.

**The question is:**

Should the new implementation offer specialized `PRAGMA` statements for cipher configuration in addition to the C interface, user-defined functions in `SELECT` statements, and URI parameters? Or would that add too much confusion?

Please share your opinions! Thanks.

**Comment by utelle** [comment]
Compared to `SELECT` statements calling a user-defined function `PRAGMA` statements with the same functionality add in fact only syntactic sugar. However, from my point of view `PRAGMA` statements are more intuitive than `SELECT` statements.

I found out that intercepting the internal handling of `PRAGMA` statements could be accomplished by a small and simple patch to the SQLite code (affecting only a single line of code). So, this leads again to the question asked in issue #2: **Allow patching of SQLite code? Or avoid it _at all costs_?**

**Comment by Willena** [comment]
I agree with you.
Using `PRAGMA` to configure the lib is way more intuitive than using SELECT. I'm not against patching SQLite source code, even if it would have been better for maintenance to keep SQLite source as is.

**Comment by utelle** [comment]
At least for the `PRAGMA` handling the patch would be really small. And it could be automated, so that patching would be simply the task of running a small script. (Solving other issues might be more complex, though.)

On the other hand, the syntax of `PRAGMA` statements is rather simple:
```SQL
PRAGMA [schema.]id = value;
```
The _value_ can only be a number or a string, not an expression. In this respect, `SELECT` statements are much more flexible.

A complication in the context of this project is that multiple ciphers are supported, and each cipher has configuration options. Therefore it will be necessary to make a design decision:

1) Make `PRAGMA` identifiers **unique**, i.e. `sqlcipher_kdf_iter`, `sqlcipher_kdf_iter_default`, `chacha20_kdf_iter`, `chacha20_kdf_iter_default` and so on.
2) Make `PRAGMA` identifiers **context-dependent**, i.e. `kdf_iter` and `kdf_iter_default` would affect only the currently selected cipher.
3) Provide a single configuration `PRAGMA` for each cipher, i.e.
 `PRAGMA sqlcipher_config='kdf_iter=1;hmac_use=1;...'`

Variant 1 would bring us clunky `PRAGMA` identifiers like `sqlcipher_plaintext_header_size_default`. By using the non-standard character `$` in identifiers, the components of the identifiers could be separated more visibly: `sqlcipher$plaintext_header_size$default`, but this is also rather ugly.

Variant 2 is syntactically a bit less clunky, but more error-prone. However, the current implementation of providing configuration parameters via URI database file names uses more or less this approach already. So, this might be the preferrable approach.

Variant 3 doesn't clutter the `PRAGMA` identifier name space, but it is a bit cumbersome to use, especially for querying parameter values.

**Comment by Willena** [comment]
If I have to choose from the three architecture to setup, I would prefer variant 2 or variant 3. 

Variant 2 implies to give `PRAGMA` statements in order (first set the cipher algorithm then change parameters). It can give flexibility if a new cipher algorithm is added in the future. 
An example: 
```sql
PRAGMA cipher=chacha20;
PRAGMA cipher.kdf_iter=1;
...
```

Variant 3 is very close to how we configure cipher with the URI. Why not using the same syntax ? (I have no idea about if function are already available to process each individual param of the URI)
An example :
```sql
PRAGMA cipher_config="cipher=sqlcipher&legacy=1&kdf_iter=4000";
```

Just to clarify, will the `SELECT` statement always be used to collect the current config ?

**Comment by utelle** [comment]
> If I have to choose from the three architectures to setup, I would prefer variant 2 or variant 3.

At the moment, my preference would be variant 2.

> Variant 2 implies to give `PRAGMA` statements in order (first set the cipher algorithm then change parameters). It can give flexibility if a new cipher algorithm is added in the future.

Fair comment. And in case no cipher scheme is given, the current default cipher would be used.

> An example:
> 
> ```sql
> PRAGMA cipher=chacha20;
> PRAGMA cipher.kdf_iter=1;
> ...
> ```

One remark: it is not possible to use a dot in identifiers. If you use a dot, the first component (here `cipher`) would be interpreted as a database schema name. I would simply use the configuration parameter names as `PRAGMA` identifiers.

One option for the parameter variations _default_, _min_ and _max_ (as supported by the user-defined configuration function) would be to use quotes, for example:
```SQL
PRAGMA [default:kdf_iter];
-- or
PRAGMA "min:kdf_iter";
```

> Variant 3 is very close to how we configure cipher with the URI. Why not using the same syntax ? (I have no idea about if function are already available to process each individual param of the URI)

Probably there exist internal SQLite functions to parse URI parameters. However, one has to specify cipher configuration parameters immediately on opening a database. So, if a user prefers the URI syntax, he should simply use URI parameters directly.

> Just to clarify, will the `SELECT` statement always be used to collect the current config ?

I intend to keep the option to query configuration parameters via `SELECT`. However, in addition I would offer the option to query configuration parameters via `PRAGMA`: simply leaving out the equal sign and the parameter value will do that (as with other `PRAGMA` statements.

**Comment by Willena** [comment]
> Fair comment. And in case no cipher scheme is given, the current default cipher would be used.

Nothing has changed here, default stay default. If you change the cipher, but not its parameters then its default parameters should be used.

> One remark: it is not possible to use a dot in identifiers. If you use a dot, the first component (here cipher) would be interpreted as a database schema name. I would simply use the configuration parameter names as PRAGMA identifiers.

Plus, what if I decide to call my schema "cipher"... There will be a conflict between names (cipher config or schema name)

> One option for the parameter variations default, min and max (as supported by the user-defined configuration function) would be to use quotes, for example:
>```
> PRAGMA [default:kdf_iter];
> -- or
> PRAGMA "min:kdf_iter";
> ```

`PRAGMA "min:kdf_iter";` `PRAGMA "default:kdf_iter";` `PRAGMA "kdf_iter";`   Same name as in the wxsqlite3_config() function. This is not a problem for me.

> So, if a user prefers the URI syntax, he should simply use URI parameters directly.

In that case only variant 2 is viable (and should be implemented). 

>  I would offer the option to query configuration parameters via PRAGMA

This is exactly what I had in mind 😄

**Comment by utelle** [comment]
> > One remark: it is not possible to use a dot in identifiers. If you use a dot, the first component (here cipher) would be interpreted as a database schema name. I would simply use the configuration parameter names as PRAGMA identifiers.
> 
> Plus, what if I decide to call my schema "cipher"... There will be a conflict between names (cipher config or schema name)

The syntax of the `PRAGMA` statement is

```SQL
PRAGMA [schema.]id [= value];
```
If `schema` is not specified, the schema `main` will be used. `main` and `temp` are predefined schema names for the main database and the temporary database. If you `ATTACH` further database files, you can assign any valid schema name to the attached database, even the schema name `cipher`. There will be no conflict with identifiers, because they can't have dots in them, unless you use quotes (in which case the dot would no longer be interpreted as the schema name delimiter).

> > So, if a user prefers the URI syntax, he should simply use URI parameters directly.
> 
> In that case only variant 2 is viable (and should be implemented).

`PRAGMA` support will be added in the near future.

However, there is still one issue left that I will have to decide about first: the wxSQLite3 implementation of the encryption parameter configuration does not include schema references. That is, there are 2 sets of parameter values associated with the main database: the default set and the current set. The current set is used solely for the next database connection (for example, when attaching a database). After the operation parameter values are rest to the defaults.

Now, with the `PRAGMA` approach we have the (optional) schema name. After attaching a database, one could use `PRAGMA` statements with schema name to modify configuration parameters, **before** the `PRAGMA schema.KEY=...` would be issued to establish the encrypted database connection. After the latter `PRAGMA` has been issued, further configuration `PRAGMA` statements with schema name would have no effect, except maybe for setting a configuration for a later `PRAGMA schema.REKEY=...` operation.

This leads me to the conclusion to prohibit the use of the `default:` prefix in configuration `PRAGMA` statements. Instead the behaviour would be as follows: for schema `main` the values from the default set would be queried or manipulated, for all other schemas the values from the current set would be queried or manipulated. This could impose a problem only, if one is used to set configuration parameters for databases that are going to be attached (without affecting the default values), **before** actually issuing the `ATTACH` statement.

**Comment by utelle** [comment]
Today I implemented the encryption key handling for the `ATTACH` statement.

Additionally, I updated the [documentation](https://utelle.github.io/SQLite3MultipleCiphers/). The `PRAGMA` statements that are available for configuring the cipher schemes are now documented.

From my point of view the new implementation of the encryption extension should be more or less operational. However, most likely there are a few bugs left, waiting for being found.
